kuryr_kubernetes/cni/daemon/service.py

```diff
@@ -359,13 +360,15 @@ def __init__(self, worker_id, metrics):
 
     def _start_metric_updater(self):
         while self.is_running:
-            if self.metrics:
-                pod_name = list(self.metrics.keys())[0]
-                with lockutils.lock(pod_name):
-                    labels = self.metrics[pod_name]['labels']
-                    duration = self.metrics[pod_name]['duration']
-                    self.prometheus_exporter.update_metric(labels, duration)
-                    del self.metrics[pod_name]
+            try:
+                metric = self.metrics.get(timeout=1)
+            except queue.Empty:
+                continue
+            pod_name = metric['name']
+            labels = metric['labels']
+            duration = metric['duration']
+            with lockutils.lock(pod_name):
```

Comment:
why do we need to lock here?

Reply:
Hm, maybe not. So I left it here thinking that prometheus_exporter might not be thread safe. Or that it's called from various places. But seems like it can only be called from this thread and we're guaranteed it's sequential.

Reply:
right, no strong opinion. If you believe leaving it here might be safer, it's fine.